docs/rfcs/0012-introduce-ui-store.md

```diff
@@ -0,0 +1,163 @@
+
+There are some advantages to this change:
+
+* Simplicity for `Store`s that are meant for UI facing work.
+* Unit testing can now occur on the test framework's thread.
+* Fewer resources needed for context shifting between threads[^1].
+
+For an example of thread simplicity, an `Engine` typically has its own 'engine thread' to perform async work and post/request results to the main thread (these APIs are identified with the `@UiThread` annotation). Once we get the callback for those results, we then need to dispatch an action to the store that will then happen on a `Store` thread. Feature components then observe for state changes and then make UI changes on the main thread. A simplified form of this thread context switching can be seen in the example below:
+
+```kotlin
+// engine thread
+engineView.requestApiResult { result ->
+  // received on the main thread.
+  store.dispatch(UpdateResultAction(result))
+}
+
+// store thread
+fun reduce(state: State, action: Action) {
+  is UpdateResultAction -> {
+    // do things here.
+  }
+}
+
+// store thread
+Middleware {
+  override fun invoke(
+    context: MiddlewareContext<State, Action>,
+    next: (Action) -> Unit,
+    action: Action,
+  ) {
+    // perform side-effects that also happen on the store thread.
+  }
+}
+
+// main thread
+store.flowScoped { flow ->
+  flow.collect {
+    MainScope().launch {
+      // perform work on the main thread.
+    }
+  }
+}
+```
+
+With the changes in this RFC, this switching of threads can be reduced (notable comments marked with 📝):
+
+```kotlin
+// engine thread
+engineView.requestApiResult { result ->
+  // received on the main thread.
+  store.dispatch(UpdateResultAction(result))
+}
+
+// 📝 main thread - now on the same thread, processed immediately.
+fun reduce(state: State, action: Action) {
+  is UpdateResultAction -> {
+    // do things here.
+  }
+}
+
+// 📝 main thread - now on the same thread, processed immediately.
+Middleware {
+  override fun invoke(
+    context: MiddlewareContext<State, Action>,
+    next: (Action) -> Unit,
+    action: Action,
+  ) {
+    // 📝 perform side-effects that now happen on the main thread.
+  }
+}
+
+// main thread
+store.flowScoped { flow ->
+  flow.collect {
+      // perform work on the main thread.
+    }
+  }
+}
+```
+
+Additionally, from [performance investigations already done][2], we know that Fenix creates over a hundred threads within a few seconds of startup. Reducing the number of threads for Stores that do not have a strong requirement to run on a separate thread will lower the applications memory footprint.
```

Comment:
@kycn I noted this bit here as well as it's something that'll help us make incremental wins once this lands and if you're interested in working on this.

Reply:
There are some quick wins in this list of the Fenix/AC threads: https://github.com/mozilla-mobile/fenix/blob/main/docs/List-of-fenix-threads.md

These don't include the Store created threads which we probably have quite a few on startup. Let me know if you want help in working on reducing the Store threads!

Reply:
I will just note tangentially that this specific concern would be eliminated if we moved to a full Redux, single-store model, which I would still like to investigate doing (at least in Fenix).

However, this will require quite a bit of prototyping and investigation to make sure we can create ephemeral state that goes out of scope based on the screen so we don't pollute users memory unnecessarily.

Even if we do go that route eventually, I don't foresee using `UiStore`s in the interim from adding an difficulty to that migration.

Reply:
Yeah, let's talk about a single store model separately. I have not personally seen any proposal that describes what the _concrete_ benefits our.

> > Even if we do go that route eventually, I don't foresee using UiStores in the interim from adding an difficulty to that migration.

I find this puzzling. Isn't it a bit quick to discount this while we are doing this work to make UI screen stores better? cc: @csadilek if you want to add more clarity that I might have missed in this doc or thread.